themes/default/content/blog/exploring-circular-dependencies/index.md

```diff
@@ -0,0 +1,226 @@
+
+How then could we make the process easier and clearer for users who need to resolve this kind of circular dependency regularly? We explored four options as part of the hackathon, so let’s go through each one, getting progressively closer to a possible solution.
+
+## Option 1: State
+
+Instead of referencing outputs directly from our stack’s previous deploy, and manually deploying twice, we could introduce a special type called “state”.
+
+The `state` object does 3 tasks:
+
+1. Read the value from the previous deployment, or provide a default if it’s the first deployment.
+2. Update the value in the stack state using a “set” method.
+3. Automatically triggers another deployment if its value has changed.
+
+```typescript
+const pongName = new pulumi.State("pong-name", {
+  initialValue: "",
+});
+
+const ping = new CallbackFunction("ping", {
+  callback: pingPongHandler,
+  environment: {
+    variables: {
+      OPPONENT_FN_NAME: pongName.value,
+    },
+  },
+});
+
+const pong = new CallbackFunction("pong", {
+  callback: pingPongHandler,
+  environment: {
+    variables: {
+      OPPONENT_FN_NAME: ping.name,
+    },
+  },
+});
+
+// Engine re-runs "up" if this changes the value
+pongName.set(pong.name);
+```
+
+Deployments should always match the preview, but when triggering the re-deployment, we’re making a change that wasn’t in the preview. This action could also result in a never-ending loop where every deployment makes another change to the state and triggers another deployment. So let’s explore a different option that might not have this issue.
+
+## Option 2: Refs
+
+This option looks almost identical to the “state” concept above, except we’re calling them references (refs):
+
+```typescript
+const pongName = new pulumi.Ref("pong-name", {
+  initialValue: "",
+});
+
+const ping = new CallbackFunction("ping", {
+  callback: pingPongHandler,
+  environment: {
+    variables: {
+      OPPONENT_FN_NAME: pongName, // Special type the Pulumi engine will track
+    },
+  },
+});
+
+const pong = new CallbackFunction("pong", {
+  callback: pingPongHandler,
+  environment: {
+    variables: {
+      OPPONENT_FN_NAME: ping.name,
+    },
+  },
+});
+
+// Determine affected resources and re-update
+pongName.set(pong.name);
+```
+
+The difference between refs and state is that, instead of triggering a whole re-deploy when the value changes, refs would be integrated into the dependency graph—similar to resources. When the value of the ref changes, we could internally work out which resources depend on that ref and change the target state for only those resources.
+
+A reference would only be allowed to be updated once in a program to avoid creating never-ending loops. For example, if a resource uses a reference value, then we update the reference value from that same resource; if that continues indefinitely, the program might never exit.
```

Comment:
I don't think this sort of loop is actually expressible with Refs? The whole point of refs is you can't get the updated resource outputs to be able to call Ref.Set again?

Reply:
Hmm, perhaps not - shall I just remove this paragraph then?

Reply:
Have reworked 👍